### ISSUE TITLE
How to keep track on the prediction (accuracy) of a botorch model

### ISSUE THREAD

**Issue by amgonzalezd** [facebook/Ax#1409]
Hi! I would like to keep track on the prediction accuracy of the botorch model at each trial. I've been looking for some similar issue, but i haven't found anything. When using a synthetic function I would like to know at each trial how good the model is performing, for example by calculating the RMSE between true value of the function and predicted output of the model. This would be interesting to set a stop criteria for the experiment or to check the improvement of the model at each step. 
Two questions on this:
1. Is this information kept somewhere and is there some way to access it?

2. If not, what would be the best way to do it?
I was thinking of adding a tracking_metric for it. Using the Service API I would take the model in the loop, predict the values and add it as a tracking-metric:
```
ax_client = AxClient()
ax_client.create_experiment(
    name="synthetic",
    parameters=parameters
    objectives={"TestFunction": ObjectiveProperties(minimize=True)},
    tracking_metric_names=['ModelPrediction']
)


for i in range(30):
    parameters, trial_index = ax_client.get_next_trial()
    model = ax_client.generation_strategy.model
    obs_feat = ObservationFeatures(parameters=parameters)
    try:  # Quick and dirty solution to avoid program crashing when trying to "predict" something with the sobol model, that doesn't implement this method. It should be done cleaner.
        prediction = model.predict(obs_feat)
        # Here the result is in the form of ({'TestFunction': [predicted_value]}, {'TestFunction: {'TestFunction': [squared error to metric??]}})

    except:
        prediction = None
  
    ax_client.complete_trial(trial_index=trial_index, raw_data={**evaluate(parameters), "ModelPrediction": (predicted_value, sem))

```

Does this make sense??
Thank you in advance!!

**Comment by mpolson64** [comment]
The best way to get prediction accuracy of a botorch model is via cross validation, which Ax exposes via a `cross_validate` function. The following snippet will return a list of (observation, prediction) pairs for each observation:
```
from ax.modelbridge.cross_validation import cross_validate

cv = cross_validate(model=ax_client.generation_strategy.model, folds=-1)
```
With folds specified as -1 (the default) we preform leave-one-out cross validation, where we predict once for each observation, using all other observations as a training set and using the selected observation as a single-item test set. You may use these values to calculate the RMSE if you would like, or you can plot the result using `interact_cross_validation_plotly`.

If you want to understand how the model fit improves over time you may do this after every trial as you are doing now, or use `cross_validate_by_trial(model: ModelBridge, trial: int)` to get the cross validation for all trials up to the trial index specified. 

As an aside, controlling the GenerationStrategy based on the model fit is an active area of development for the team and we're hoping to get something out for this in the future. Let me know if you have any other questions about this!

**Comment by amgonzalezd** [comment]
I've tried this out and it works wonderful! 
If there's any advances on the controlled GenerationStrategy based on the model fit, I'd be very interested to know!

Thank you very much!!

**Comment by mpolson64** [comment]
Happy this was helpful! And when advances are made on model fit dependent GenerationStrategies we will let everyone know in the patch notes and ideally a tutorial on our website :)

Closing this issue, please feel free to continue asking questions in new issues.
